bi_view_editor/models/ir_model.py

```diff
@@ -0,0 +1,291 @@
+                               field['model_id']) for field in field_data])
+            return model_ids
+
+        def _get_join_nodes_dict(model_ids, new_field):
+            join_nodes = []
+            for alias, model_id in model_ids.items():
+                if model_id == new_field['model_id']:
+                    join_nodes.append({'table_alias': alias})
+            for field in self.get_related_fields(model_ids):
+                c = [field['join_node'] == -1, field['table_alias'] == -1]
+                a = (new_field['model'] == field['relation'])
+                b = (new_field['model_id'] == field['model_id'])
+                if (a and c[0]) or (b and c[1]):
+                    join_nodes.append(field)
+            return join_nodes
+
+        def remove_duplicate_nodes(join_nodes):
+            seen = set()
+            nodes_list = []
+            for node in join_nodes:
+                node_tuple = tuple(node.items())
+                if node_tuple not in seen:
+                    seen.add(node_tuple)
+                    nodes_list.append(node)
+            return nodes_list
+
+        model_ids = _get_model_ids(field_data)
+        keys = [(field['table_alias'], field['id'])
+                for field in field_data if field.get('join_node', -1) != -1]
+        join_nodes = _get_join_nodes_dict(model_ids, new_field)
+        join_nodes = remove_duplicate_nodes(join_nodes)
+
+        return list(filter(
+            lambda x: 'id' not in x or
+                      (x['table_alias'], x['id']) not in keys, join_nodes))
+
+    @api.model
+    def get_fields(self, model_id):
+        domain = [
+            ('model_id', '=', model_id),
+            ('store', '=', True),
+            ('name', 'not in', NO_BI_FIELDS),
+            ('ttype', 'not in', NO_BI_TTYPES)
+        ]
+        fields_dict = []
+        filtered_fields = self._search_fields(domain)
+        for field in filtered_fields:
+            fields_dict.append(
+                {'id': field.id,
+                 'model_id': model_id,
+                 'name': field.name,
+                 'description': field.field_description,
+                 'type': field.ttype,
+                 'custom': False,
+                 'model': field.model_id.model,
+                 'model_name': field.model_id.name
+                 }
+            )
+        sorted_fields = sorted(
+            fields_dict,
+            key=lambda x: x['description'],
+            reverse=True
+        )
+        return sorted_fields
+
+    @api.model
+    def create(self, vals):
+        if self._context and self._context.get('bve'):
+            vals['state'] = 'base'
+        res = super(IrModel, self).create(vals)
+
+        # this sql update is necessary since a write method here would
+        # be not working (an orm constraint is restricting the modification
+        # of the state field while updating ir.model)
+        q = "UPDATE ir_model SET state = 'manual' WHERE id = %s"
```

Comment:
I don't understand. Why putting `vals['state'] = 'base' `in https://github.com/OCA/reporting-engine/pull/202/files#diff-5d42ae09ec3aa29081fdaefd4d438201R274 and then put this query? Will all modules have `state = 'manual'?` Then why in https://github.com/OCA/reporting-engine/pull/202/files#diff-824d3c4f48b7f296e915d38877993ceaR390 there is the `'state' = manual`?

Reply:
@mreficent all such strange pieces of code you are finding out are used as workarounds, in order to bypass the standard Odoo constraints ...

Reply:
But I want to full understand them. Because maybe we can think out other workarounds less intrusive.

Reply:
If we keep state =  'manual' here, than a constraint of standard Odoo is triggered while invoking the create. In order to make the create() working fine, the workaround consists in:

- set state =  '< something >' (something different than 'manual' or 'auto')
- invoke the create()
- set state back to 'manual', using SQL to bypass anther ORM contraint

I think the best way to understand this is to play with it, commenting out those parts of code and debugging a bit. I would recommend you to do so, hopefully you will find a better way to do bypass the constraints. Unfortunately I spent so much time on that part of code that I give up to continue to search a better solution.

Reply:
Ok. I will play with it :)